sw/host/tests/chip/sysrst_ctrl/sysrst_ctrl_ec_rst_l.rs

```diff
@@ -0,0 +1,232 @@
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::test_utils::test_status::TestStatus;
+use opentitanlib::uart::console::UartConsole;
+use opentitanlib::{collection, execute_test};
+
+use sysrst_ctrl::{Config, setup_pins, set_pins, read_pins};
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    /// Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "10s")]
+    timeout: Duration,
+}
+
+struct Params<'a> {
+    opts: &'a Opts,
+    transport: &'a TransportWrapper,
+    uart: &'a dyn Uart,
+    config: &'a Config,
+}
+
+static CONFIG: Lazy<HashMap<&'static str, Config>> = Lazy::new(|| {
+    collection! {
+        "hyper310" => Config {
+            // key0_in, key1_in
+            output_pins: vec!["IOR6", "IOR7"],
+            open_drain: vec![false, false],
+            // ec_rst, flash_wp
+            input_pins: vec!["IOR8", "IOR9"],
+            // ec_rst, flash_wp
+            pullup_pins: vec!["IOR8", "IOR9"],
+        },
+    }
+});
+
+const COMBO_PINS_HIGH: u8 = 0x3;
+const COMBO_PINS_LOW: u8 = 0;
+const COMBO_PINS_KEY0_HIGH: u8 = 1;
+const EC_RST_FLASH_WP_HIGH: u8 = 0x3;
+const EC_RST_PIN_INDEX: u8 = 0;
+const FLASH_WP_PIN_INDEX: u8 = 1;
+const EC_RST_PIN: &str = "IOR8";
+
+const EC_RST_COMBO_PULSE_WIDTH_US: u64 = 100_000; // Set on the device.
+const EC_RST_STRETCH_PULSE_WIDTH_US: u64 = 200_000; // Set on the device.
+const EC_RST_PULE_WIDTH_MARGIN_US: u64 = 5_000;
+
+fn sync_with_sw(params: &Params) -> Result<()> {
+    UartConsole::wait_for(params.uart, &TestStatus::InWfi.wait_pattern(), params.opts.timeout)?;
+    Ok(())
+}
+
+fn chip_sw_sysrst_ctrl_input(
+    params: &Params,
+) -> Result<()> {
+
+    // Setup transport pins.
+    setup_pins(params.transport, params.config)?;
+    // Set combo pins to high.
+    set_pins(params.transport, params.config, COMBO_PINS_HIGH)?;
+
+    // Reset target now so that we can be sure that the pins are in the right
+    // configuration before running the test.
+    params.transport.reset_target(params.opts.init.bootstrap.options.reset_delay, true)?;
+
+    // Wait until device has setup pins and is waiting for combo.
+    sync_with_sw(params)?;
+    // On POR, the device releases the EC_RST and FLASH_WP pins.
+    ensure!(read_pins(params.transport, params.config)? == EC_RST_FLASH_WP_HIGH);
+
+    // We are going to set the combo pins to low and monitor the GPIO.
+    let gpio_monitoring = params.transport.gpio_monitoring()?;
+    let clock_nature = gpio_monitoring.get_clock_nature()?;
+    let ClockNature::Wallclock { resolution, .. } = clock_nature else {
+        bail!("Cannot measure EC reset pulse width: the transport GPIO monitor does not have reliable clock source");
+    };
+    log::info!("GPIO monitor clock resolution: {resolution}");
+    // Convert a timestamp in transport unit to something in nanoseconds.
+    let timestamp_to_ns = |timestamp: u64| -> u64 {
```

Comment:
Should this closure be defined closer to where it's  used?

Reply:
My intention was to define it next to where `resolution` is defined because it's meaning is not super clear. But I wil move it.